File: server/app/routes/oauth.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.services.oauth_service import OAuthService

oauth_bp = Blueprint('oauth', __name__)

@oauth_bp.route('/github/connect', methods=['GET'])
@jwt_required()
def connect_github():
    url, error = OAuthService.get_github_auth_url()
    if error:
        return jsonify({'error': error}), 400
    return jsonify({'url': url}), 200

@oauth_bp.route('/github/callback', methods=['POST'])
@jwt_required()
def github_callback():
    data = request.get_json()
    code = data.get('code')
    if not code:
        return jsonify({'error': 'Code is required'}), 400
    
    current_user_id = get_jwt_identity()
    oauth, error = OAuthService.handle_github_callback(code, current_user_id)
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'message': 'GitHub connected successfully',
        'data': oauth.to_dict()
    }), 200

@oauth_bp.route('/github/disconnect', methods=['DELETE'])
@jwt_required()
def disconnect_github():
    current_user_id = get_jwt_identity()
    success, error = OAuthService.disconnect_github(current_user_id)
    
    if not success:
        return jsonify({'error': error}), 400
    
    return jsonify({'message': 'GitHub disconnected successfully'}), 200

@oauth_bp.route('/linkedin/connect', methods=['GET'])
@jwt_required()
def connect_linkedin():
    url, error = OAuthService.get_linkedin_auth_url()
    if error:
        return jsonify({'error': error}), 400
    return jsonify({'url': url}), 200

@oauth_bp.route('/linkedin/callback', methods=['POST'])
@jwt_required()
def linkedin_callback():
    data = request.get_json()
    code = data.get('code')
    if not code:
        return jsonify({'error': 'Code is required'}), 400
    
    current_user_id = get_jwt_identity()
    oauth, error = OAuthService.handle_linkedin_callback(code, current_user_id)
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'message': 'LinkedIn connected successfully',
        'data': oauth.to_dict()
    }), 200

@oauth_bp.route('/linkedin/disconnect', methods=['DELETE'])
@jwt_required()
def disconnect_linkedin():
    current_user_id = get_jwt_identity()
    success, error = OAuthService.disconnect_linkedin(current_user_id)
    
    if not success:
        return jsonify({'error': error}), 400
    
    return jsonify({'message': 'LinkedIn disconnected successfully'}), 200

File: client/kl-hackathon/src/functions/oauth.js
import { API_BASE_URL } from '../config';

const API_URL = API_BASE_URL;

const handleResponse = async (response, errorMessage) => {
    const data = await response.json();
    if (!response.ok) {
        throw new Error(data.error || errorMessage);
    }
    return data;
};

export const connectGithub = async (token) => {
    try {
        const response = await fetch(`${API_URL}/oauth/github/connect`, {
            method: 'GET',
            headers: {
                Authorization: `Bearer ${token}`
            }
        });
        return handleResponse(response, 'Failed to initiate GitHub connection');
    } catch (error) {
        throw error;
    }
};

export const handleGithubCallback = async (token, code) => {
    try {
        const response = await fetch(`${API_URL}/oauth/github/callback`, {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ code })
        });
        return handleResponse(response, 'Failed to connect GitHub account');
    } catch (error) {
        throw error;
    }
};

export const disconnectGithub = async (token) => {
    try {
        const response = await fetch(`${API_URL}/oauth/github/disconnect`, {
            method: 'DELETE',
            headers: {
                Authorization: `Bearer ${token}`
            }
        });
        return handleResponse(response, 'Failed to disconnect GitHub account');
    } catch (error) {
        throw error;
    }
};

export const connectLinkedin = async (token) => {
    try {
        const response = await fetch(`${API_URL}/oauth/linkedin/connect`, {
            method: 'GET',
            headers: {
                Authorization: `Bearer ${token}`
            }
        });
        return handleResponse(response, 'Failed to initiate LinkedIn connection');
    } catch (error) {
        throw error;
    }
};

export const handleLinkedinCallback = async (token, code) => {
    try {
        const response = await fetch(`${API_URL}/oauth/linkedin/callback`, {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ code })
        });
        return handleResponse(response, 'Failed to connect LinkedIn account');
    } catch (error) {
        throw error;
    }
};

export const disconnectLinkedin = async (token) => {
    try {
        const response = await fetch(`${API_URL}/oauth/linkedin/disconnect`, {
            method: 'DELETE',
            headers: {
                Authorization: `Bearer ${token}`
            }
        });
        return handleResponse(response, 'Failed to disconnect LinkedIn account');
    } catch (error) {
        throw error;
    }
};
